src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -584,6 +584,17 @@ async Task<FinalCompilationTrackerState> FinalizeCompilationWorkerAsync(InProgre
                         staleCompilationWithGeneratedDocuments,
                         cancellationToken).ConfigureAwait(false);
 
+                    // Our generated documents are up to date if we just created them.  Note: when in balanced mode, we
+                    // will then change our creation policy below to DoNotCreate.  This means that any successive forks
+                    // will move us to an in-progress-state that is not running generators.  And the next time we get
+                    // here and produce a final compilation, this will then be 'false' since we'll be reusing old
+                    // generated docs.
+                    //
+                    // This flag can then be used later when we hear about external user events (like save/build) to
+                    // decide if we need to do anything.  If the generated documents are up to date, then we don't need
+                    // to do anything in that case.
+                    var generatedDocumentsUpToDate = creationPolicy.GeneratedDocumentCreationPolicy == GeneratedDocumentCreationPolicy.Create;
```

Comment:
note: looking at the code below this line is very relevant.

Reply:
importantly, we only generate the docs when the policy is `GeneratedDocumentCreationPolicy.Create`, but in balanced mode we then immediately transition after this line to .DoNotCreate mode.

So the final state will say "i just generated up to date docs, but any forks off of me should reuse the docs and not regenerate them".  Those forks will then end up back in the FinalCompilationTrackerState, except this time they will say "i did *not* just generate the SG docs" as their creation policy on this line was DoNotCreate.